From: Boyuan Yang <073plan@gmail.com>
Date: Mon, 5 Dec 2016 16:21:51 +0800
Subject: Force utf8 for retrieved window name from Gtk2.
Bug-Debian: https://bugs.debian.org/847051
Bug: https://bugs.launchpad.net/ubuntu/+source/shutter/+bug/1556021
Last-Update: 2016-12-05

Transform the window name retrieved from Gtk2 to utf8
to prevent messy encoding problem. Otherwise latin-1
would be used and cause mess.
---
 bin/shutter | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/bin/shutter
+++ b/bin/shutter
@@ -30,6 +30,11 @@
 use strict;
 use warnings;
 
+#Deal with encoding problem
+use 5.12.0;
+use Encode qw/is_utf8 decode/;
+use Encode::X11;
+
 #Gnome2 libraries
 use Gnome2;
 use Gnome2::Wnck;
@@ -8624,7 +8629,14 @@
         #add all windows to menu to capture it directly
         foreach my $win ( $wnck_screen->get_windows_stacked ) {
             if ( $active_workspace && $win->is_on_workspace($active_workspace) ) {
-                my $window_item = Gtk2::ImageMenuItem->new_with_label( $win->get_name );
+                my $raw_name = $win->get_name;
+                my $name;
+                eval { $name = decode( 'UTF-8' , $raw_name, 1); };
+                # fallback if utf-8 decoding fails.
+                # See https://mail.gnome.org/archives/gtk-perl-list/2016-December/msg00004.html
+                $name = decode( 'x11-compound-text', $raw_name ) if $@;
+
+                my $window_item = Gtk2::ImageMenuItem->new_with_label( $name );
                 foreach my $child ( $window_item->get_children ) {
                     if ( $child =~ /Gtk2::AccelLabel/ ) {
                         $child->set_width_chars(50);
